File: packages/kensho-cucumber-jvm/src/main/java/com/kaizenreport/kensho/Kensho.java
package com.kaizenreport.kensho;

import java.io.File;
import java.nio.file.Path;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Public Kensho helper API. Available to test code regardless of which adapter (junit5, testng,
 * cucumber-jvm) is registered as the listener — they all populate the same thread-local
 * {@link KenshoContext}.
 *
 * <p>All methods are no-ops when called outside a running test, so it's safe to sprinkle them in
 * shared utility code.
 *
 * <pre>{@code
 * import com.kaizenreport.kensho.Kensho;
 *
 * @Test
 * void login() {
 *   try (Kensho.Step s = Kensho.step("open the login page")) {
 *     // ...
 *   }
 *   Kensho.label("team", "growth");
 *   Kensho.link("https://jira.example.com/browse/PROJ-123", "jira", "PROJ-123");
 *   Kensho.attach(new File("/tmp/login.png"), "screenshot");
 * }
 * }</pre>
 */
public final class Kensho {

  private Kensho() {}

  /** Open a new step. Use as an {@link AutoCloseable} in a try-with-resources block. */
  public static Step step(String title) {
    return step(title, null);
  }

  public static Step step(String title, String action) {
    CaseScratch scratch = KenshoContext.current();
    if (scratch == null) return Step.NOOP;
    Map<String, Object> stepObj = new LinkedHashMap<>();
    stepObj.put("id", "step_" + shortId());
    stepObj.put("title", title == null ? "" : title);
    stepObj.put("status", "pass");
    stepObj.put("startedAt", Instant.now().toString());
    stepObj.put("duration", 0);
    if (action != null && !action.isEmpty()) stepObj.put("action", action);

    Map<String, Object> parent = scratch.stepStack.peek();
    if (parent != null) {
      @SuppressWarnings("unchecked")
      java.util.List<Map<String, Object>> children =
          (java.util.List<Map<String, Object>>) parent.get("children");
      if (children == null) {
        children = new java.util.ArrayList<>();
        parent.put("children", children);
      }
      children.add(stepObj);
    } else {
      scratch.steps.add(stepObj);
    }
    scratch.stepStack.push(stepObj);
    return new Step(scratch, stepObj, System.nanoTime());
  }

  /** Set {@code case.labels[key] = value}. No-op outside an active test. */
  public static void label(String key, String value) {
    CaseScratch scratch = KenshoContext.current();
    if (scratch == null || key == null || key.isEmpty()) return;
    scratch.labels.put(key, value == null ? "" : value);
  }

  /** Attach a hyperlink to the running case. */
  public static void link(String url) {
    link(url, null, null);
  }

  public static void link(String url, String kind, String label) {
    CaseScratch scratch = KenshoContext.current();
    if (scratch == null || url == null || url.isEmpty()) return;
    Map<String, String> entry = new LinkedHashMap<>();
    entry.put("url", url);
    if (kind != null && !kind.isEmpty()) entry.put("kind", kind);
    if (label != null && !label.isEmpty()) entry.put("label", label);
    scratch.links.add(entry);
  }

  /**
   * Copy a file into the run's {@code attachments/<caseId>/} folder and register it on the case
   * (or on the innermost open step). No-op outside a test or when the file is missing.
   */
  public static Map<String, Object> attach(File file, String kind) {
    return attach(file == null ? null : file.toPath(), kind, null, null);
  }

  public static Map<String, Object> attach(Path path, String kind) {
    return attach(path, kind, null, null);
  }

  public static Map<String, Object> attach(Path path, String kind, String name, String mimeType) {
    CaseScratch scratch = KenshoContext.current();
    if (scratch == null || path == null) return null;
    KenshoContext.AttachmentWriter w = KenshoContext.writer();
    if (w == null) return null;
    Map<String, Object> record = w.register(scratch, path, kind, name, mimeType);
    if (record == null) return null;
    Map<String, Object> innerStep = scratch.stepStack.peek();
    if (innerStep != null) {
      @SuppressWarnings("unchecked")
      java.util.List<Map<String, Object>> arr =
          (java.util.List<Map<String, Object>>) innerStep.get("attachments");
      if (arr == null) {
        arr = new java.util.ArrayList<>();
        innerStep.put("attachments", arr);
      }
      arr.add(record);
    } else {
      scratch.attachments.add(record);
    }
    return record;
  }

  /** Stable id of the test currently running, or {@code null}. */
  public static String currentCaseId() {
    CaseScratch scratch = KenshoContext.current();
    return scratch == null ? null : scratch.caseId;
  }

  private static String shortId() {
    return UUID.randomUUID().toString().replace("-", "").substring(0, 10);
  }

  /** Auto-closable handle returned by {@link Kensho#step}. */
  public static final class Step implements AutoCloseable {

    static final Step NOOP = new Step(null, null, 0L);

    private final CaseScratch scratch;
    private final Map<String, Object> obj;
    private final long startedNanos;
    private boolean closed;
    private String forcedStatus;

    Step(CaseScratch scratch, Map<String, Object> obj, long startedNanos) {
      this.scratch = scratch;
      this.obj = obj;
      this.startedNanos = startedNanos;
    }

    /** Mark the step as failed. Useful when an assertion fails but no exception is thrown. */
    public void fail() {
      forcedStatus = "fail";
    }

    /** Mark the step as skipped. */
    public void skip() {
      forcedStatus = "skip";
    }

    /** Add a parameter to the step ({@code parameters[]}). */
    public Step parameter(String name, String value) {
      if (obj == null || name == null) return this;
      @SuppressWarnings("unchecked")
      java.util.List<Map<String, String>> params =
          (java.util.List<Map<String, String>>) obj.get("parameters");
      if (params == null) {
        params = new java.util.ArrayList<>();
        obj.put("parameters", params);
      }
      Map<String, String> p = new LinkedHashMap<>();
      p.put("name", name);
      p.put("value", value == null ? "" : value);
      params.add(p);
      return this;
    }

    @Override
    public void close() {
      if (closed) return;
      closed = true;
      if (obj == null || scratch == null) return;
      long ms = Math.max(0L, (System.nanoTime() - startedNanos) / 1_000_000L);
      obj.put("duration", ms);
      if (forcedStatus != null) obj.put("status", forcedStatus);
      // Pop only if we're still on top — defensive against bad nesting.
      if (!scratch.stepStack.isEmpty() && scratch.stepStack.peek() == obj) {
        scratch.stepStack.pop();
      }
    }
  }
}

File: packages/dotnet-core/src/Kensho.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;

namespace KaizenReport.Kensho.Core;

// Public helper API users call from inside test bodies:
//
//     using (Kensho.Step("open the login page")) { /* ... */ }
//     Kensho.Attach("/tmp/screen.png", kind: "screenshot");
//     Kensho.Label("team", "growth");
//     Kensho.Link("https://jira.example.com/browse/PROJ-123", kind: "jira");
//
// All four are no-ops outside a running test so it's safe to call them
// from shared utilities. The adapter listeners install/clear the current
// scratch via Kensho.SetCurrent(...) on test start/end.
public static class Kensho
{
    private static readonly AsyncLocal<CaseScratch?> Slot = new();

    public static CaseScratch? Current => Slot.Value;

    public static void SetCurrent(CaseScratch? scratch) => Slot.Value = scratch;

    public static string? CurrentCaseId => Slot.Value?.CaseId;

    public static IDisposable Step(string title, string? action = null)
    {
        var scratch = Slot.Value;
        if (scratch == null) return NoopStep.Instance;
        return new ActiveStep(scratch, title, action);
    }

    public static KenshoAttachment? Attach(string path, string? kind = null, string? name = null, string? mimeType = null)
    {
        var scratch = Slot.Value;
        if (scratch == null || scratch.Copier == null) return null;
        var att = scratch.Copier(scratch, path, kind, name, mimeType);
        if (att == null) return null;

        if (scratch.StepStack.Count > 0)
        {
            var top = scratch.StepStack.Peek();
            (top.Attachments ??= new List<KenshoAttachment>()).Add(att);
        }
        else
        {
            scratch.Attachments.Add(att);
        }
        return att;
    }

    public static void Label(string key, string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(key)) return;
        scratch.Labels[key] = value ?? string.Empty;
    }

    public static void Link(string url, string? kind = null, string? label = null)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(url)) return;
        var entry = new KenshoLink { Url = url };
        if (!string.IsNullOrEmpty(kind)) entry.Kind = kind;
        if (!string.IsNullOrEmpty(label)) entry.Label = label;
        scratch.Links.Add(entry);
    }

    public static void Log(string message, string level = "info")
    {
        var scratch = Slot.Value;
        if (scratch == null) return;
        var t = Math.Max(0, DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - scratch.StartedAtMs);
        scratch.Logs.Add(new KenshoLog { T = t, Level = level, Msg = message });
    }

    // ---- Behavior (epic / feature / story) -------------------------------
    // These set the BDD behavior tree the viewer groups by, and also mirror
    // into labels so consumers reading labels{} stay in sync. Story maps to
    // behavior.scenario (the schema's leaf node) but label key "story".

    public static void Epic(string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(value)) return;
        (scratch.Behavior ??= new KenshoBehavior()).Epic = value;
        scratch.Labels["epic"] = value;
    }

    public static void Feature(string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(value)) return;
        (scratch.Behavior ??= new KenshoBehavior()).Feature = value;
        scratch.Labels["feature"] = value;
    }

    public static void Story(string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(value)) return;
        (scratch.Behavior ??= new KenshoBehavior()).Scenario = value;
        scratch.Labels["story"] = value;
    }

    // ---- Classification --------------------------------------------------

    // Severity: validated against the schema enum; unknown values are ignored
    // (no-op) rather than written, so the run.json stays valid.
    public static void Severity(string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(value)) return;
        var lo = value.Trim().ToLowerInvariant();
        if (Array.IndexOf(KenshoSchema.Severity, lo) < 0) return;
        scratch.Severity = lo;
    }

    public static void Owner(string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(value)) return;
        scratch.Owner = value;
    }

    public static void Description(string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || value == null) return;
        scratch.Description = value;
    }

    public static void Tag(string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(value)) return;
        scratch.Tags.Add(value);
    }

    // ---- Parameters ------------------------------------------------------
    // Free-form name/value parameter (no kind — distinct from adapter-captured
    // [TestCase] arguments which carry kind:"argument").
    public static void Parameter(string name, string value)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(name)) return;
        scratch.Parameters.Add(new KenshoParameter
        {
            Name = name,
            Value = value ?? string.Empty,
        });
    }

    // ---- Links -----------------------------------------------------------
    // Convenience link helpers with canonical kinds. The general Kensho.Link()
    // above stays for arbitrary kinds.

    public static void JiraLink(string idOrUrl, string? label = null)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(idOrUrl)) return;
        scratch.Links.Add(new KenshoLink
        {
            Url = idOrUrl,
            Kind = "issue",
            Label = string.IsNullOrEmpty(label) ? null : label,
        });
    }

    public static void ReferenceLink(string url, string? label = null)
    {
        var scratch = Slot.Value;
        if (scratch == null || string.IsNullOrEmpty(url)) return;
        scratch.Links.Add(new KenshoLink
        {
            Url = url,
            Kind = "reference",
            Label = string.IsNullOrEmpty(label) ? null : label,
        });
    }

    // ---- Markers ---------------------------------------------------------

    // Flaky: flag the case as known-flaky. The listeners copy this onto
    // case.flaky so the viewer renders a flaky badge.
    public static void Flaky()
    {
        var scratch = Slot.Value;
        if (scratch == null) return;
        scratch.Flaky = true;
    }

    // Muted: known failure that shouldn't count against the pass gate.
    public static void Muted()
    {
        var scratch = Slot.Value;
        if (scratch == null) return;
        scratch.Muted = true;
    }

    // KnownIssue: mute the case and attach an issue link pointing at the
    // ticket, in one call.
    public static void KnownIssue(string idOrUrl, string? label = null)
    {
        var scratch = Slot.Value;
        if (scratch == null) return;
        scratch.Muted = true;
        if (!string.IsNullOrEmpty(idOrUrl))
        {
            scratch.Links.Add(new KenshoLink
            {
                Url = idOrUrl,
                Kind = "issue",
                Label = string.IsNullOrEmpty(label) ? null : label,
            });
        }
    }

    private sealed class NoopStep : IDisposable
    {
        public static readonly NoopStep Instance = new();
        public void Dispose() { }
    }

    private sealed class ActiveStep : IDisposable
    {
        private readonly CaseScratch _scratch;
        private readonly KenshoStep _step;
        private readonly Stopwatch _sw;
        private bool _disposed;

        public ActiveStep(CaseScratch scratch, string title, string? action)
        {
            _scratch = scratch;
            _sw = Stopwatch.StartNew();
            _step = new KenshoStep
            {
                Id = "step_" + Guid.NewGuid().ToString("N").Substring(0, 10),
                Title = title ?? string.Empty,
                Status = "pass",
                StartedAt = DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ"),
            };
            if (!string.IsNullOrEmpty(action))
            {
                // We don't have an "Action" property on KenshoStep in the POCO
                // (the schema field exists; we keep the POCO lean). If callers
                // need it, we promote action into the title prefix to stay
                // legal against the schema.
                _step.Title = $"{action}: {_step.Title}";
            }

            if (_scratch.StepStack.Count > 0)
            {
                var parent = _scratch.StepStack.Peek();
                (parent.Children ??= new List<KenshoStep>()).Add(_step);
            }
            else
            {
                _scratch.Steps.Add(_step);
            }
            _scratch.StepStack.Push(_step);
        }

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;
            _sw.Stop();
            _step.Duration = _sw.ElapsedMilliseconds;
            // If the user didn't explicitly mark it failed, "pass" stays.
            // Adapters that wrap a step around an exception should set
            // status before disposal.
            if (_scratch.StepStack.Count > 0 && ReferenceEquals(_scratch.StepStack.Peek(), _step))
            {
                _scratch.StepStack.Pop();
            }
        }

        // Allow `using var s = Kensho.Step(...); s.Fail();` style if the
        // user wants to capture exception status without rethrowing.
        public ActiveStep Fail()
        {
            _step.Status = "fail";
            return this;
        }
    }
}
